### PR TITLE
refactor(schema): Enhanced "loose" collections

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- When parsing JSON data, it's useful to filter out invalid record/array elements while retaining the whole rest of well-formed data
- `LooseArray` and `LooseRecord` are helpers that allow easy creation of such schemas
- The callback argument has the similar signature as Zod's `.catch(...)` and may be used for logging
- `looseArray` and `looseRecord` usage was updated for existing schemas
- `looseValue` was removed since it doens't provide enough value

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/packagist/schema.ts
Comment (viceice) [inline lib/modules/datasource/packagist/schema.ts:174]: catch here?

File: lib/modules/datasource/packagist/schema.ts
Comment (viceice) [inline lib/modules/datasource/packagist/schema.ts:186]: catch here?

File: lib/util/schema-utils.ts
Comment (viceice) [inline lib/util/schema-utils.ts:22]: why uppercase function name?

Comment (zharinov) [reply lib/util/schema-utils.ts:22]: I think it communicates better the intent of being like a generic, but for schema

File: lib/modules/datasource/packagist/schema.ts
Comment (zharinov) [reply lib/modules/datasource/packagist/schema.ts:174]: Otherwise the whole `PackagesResponse` will fail on empty `packages` field. Previous implementation did this automatically, but it's not always the desired behavior, so I've just ensured each occurrence of older implementation has its catch value.

File: lib/util/schema-utils.ts
Comment (viceice) [reply lib/util/schema-utils.ts:24]: 🤔 @rarkins @JamieMagee @secustor WDYT?

**Review by viceice — COMMENTED** [review]
Waiting for feedback of other maintainer

File: lib/util/schema-utils.ts
Comment (secustor) [reply lib/util/schema-utils.ts:24]: I personally would expect lowercase, but IMO this comes down to personal taste.

**Comment by zharinov** [comment]
Please let's merge this

**Review by viceice — APPROVED** [review]

**Pull request merged**
